### PR TITLE
[Prototype] Place shots on `SampleMeasruement`

### PR DESCRIPTION
**Pull request by albi3ro**
This PR explores the changes necessary to place `shot` information on `SampleMeasurement` instead of the device.

So far, it seems rather straightforward.  The only things that needed editing where:
* `SampleMeasurement`
* Some refactoring changes to `MeasurementProcess`
* Transforms that act on measurement processes
    - `hamiltonian_expand`
    - `sum_expand`
    - `split_non_commuting`

Shot information was then able to carry all the way through the workflow and hit the device still on the measurement process.

Basically, the only objects affected this change would be sample measurements, transformations of sample measurements, and things that consume sample measurements.

### CODE REVIEW

**Comment by albi3ro** [comment]
We're not proceeding this way. Closing the issue.
